File: scala/problem4.scala
/*
https://projecteuler.net/problem=4

Largest palindrome product
*/

object problem4 {
  private def isPalindrome(n: Int): Boolean = {
    val s = n.toString
    s == s.reverse
  }

  def baselineAnswer: Long = {
    val best = (100 to 999)
      .flatMap(a => (100 to 999).map(b => a * b))
      .filter(isPalindrome)
      .max
    best.toLong
  }

  def optimizedAnswer: Long = {
    var best = 0
    var a = 999

    while (a >= 100) {
      val (bStart, bStep) = if (a % 11 == 0) (999, 1) else (990, 11)
      var b = bStart

      while (b >= a) {
        val prod = a * b
        if (prod <= best) {
          b = -1
        } else {
          if (isPalindrome(prod)) {
            best = prod
          }
          b -= bStep
        }
      }

      a -= 1
    }

    best.toLong
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 4")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem4.cpp
/*
https://projecteuler.net/problem=4

Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009.

Find the largest palindrome made from the product of two 3-digit numbers.

*/

#include <chrono>
#include <iostream>
using namespace std;

struct result {
    int value;
    long long time_ns;
};

bool is_palindrome(int n) {
    int reversed = 0;
    int temp = n;
    while (temp > 0) {
        reversed = reversed * 10 + temp % 10;
        temp /= 10;
    }
    return n == reversed;
}

int evaluate_baseline() {
    int max_palindrome = 0;
    for (int i = 100; i <= 999; ++i) {
        for (int j = i; j <= 999; ++j) {
            int product = i * j;
            if (product > max_palindrome && is_palindrome(product)) {
                max_palindrome = product;
            }
        }
    }
    return max_palindrome;
}

int evaluate_optimized() {
    int max_palindrome = 0;

    for (int i = 999; i >= 100; --i) {
        // Prune outer loop: with this i, the biggest possible product is i * 999.
        // If even that cannot beat the best palindrome we already have, smaller i
        // values cannot help either, so we can stop completely.
        if (i * 999 < max_palindrome) {
            break;
        }

        int j_start = 999;
        int step = 1;

        // For 6-digit palindromes abccba:
        //   abccba = 100001a + 10010b + 1100c = 11(9091a + 910b + 100c).
        // So any 6-digit palindrome is divisible by 11, meaning at least one
        // factor must be a multiple of 11. If i is not divisible by 11, force
        // j to move through multiples of 11 only.
        if (i % 11 != 0) {
            j_start = 990;
            step = 11;
        }

        for (int j = j_start; j >= i; j -= step) {
            int product = i * j;
            // Prune inner loop: j only decreases from here, so products only get
            // smaller. Once product <= current best, no later j can improve it.
            if (product <= max_palindrome) {
                break;
            }
            if (is_palindrome(product)) {
                max_palindrome = product;
            }
        }
    }

    return max_palindrome;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    int value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    int value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 4" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
